### ISSUE TITLE
Better error message when using GitLab Deploy Keys but they're not enabled for that repo

### ISSUE THREAD

**Issue by mattparkes** [argoproj/argo-cd#4441]
# Summary

When first adding a (self-hosted) GitLab repository to argocd with `argocd repo add` using a [Public Deploy Key](https://docs.gitlab.com/ee/user/project/deploy_keys/#public-deploy-keys), I forget to enable the Deploy Key in that project and got this rather unhelpful error:
```
FATA[0000] rpc error: code = Unknown desc = unknown error: remote:
```

Eventually, I used a combination of `ssh -v` and `git clone` to troubleshoot and realised that I was getting an access denied error because the Deploy Key wasn't enabled.

I feel like this is a _really_ common code path and warrants a dedicated check and error message, telling me that authentication has failed, as that would have steered me towards the correct solution faster.

# Motivation

I was setting up Argo for the first time and the error message was extremely unhelpful, causing me to waste a lot of time checking every facet (e.g connectivity, etc), rather than straight up being told auth had failed.

# Proposal

A `FATA` log line telling me that Auth had failed.

**Comment by jannfis** [comment]
I agree that this error message isn't particularly helpful at all, and I believe there are quite a few more places that provide an experience similar to this one. 

Thanks for reporting.

**Comment by spirosoik** [comment]
I can get this

**Comment by spirosoik** [comment]
@mattparkes do I miss something? could you give more details? This is what I see if I don't provide a deploy key

```
FATA[0001] rpc error: code = Unknown desc = ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
```

**Comment by mattparkes** [comment]
I'm not really sure sorry, we never went ahead with ArgoCD (just because of other priorities) and so I can't easily verify this.

I feel _reasonably_ confident that my original report was correct, so perhaps something else has changed to surface the correct error since my issue was opened.

Happy for this to be closed if that's the case! :)

**Comment by spirosoik** [comment]
@jannfis I think we can close this.

**Comment by varesa** [comment]
On ArgoCD v2.1.2+unknown (packaged by Red Hat as OpenShift GitOps) I got the following status/error message:
```
status:
  conditions:
  - lastTransitionTime: "2021-11-24T13:15:07Z"
    message: 'rpc error: code = Unknown desc = unknown error: remote: '
    type: ComparisonError
```
when I had forgotten to enable a deploy key on the git repository. It was actually this issue that I found via Google that made me realize my mistake

**Comment by DanielB1990** [comment]
Same here as for Varesa, thanks to Alexander Dalloz on the [ArgoCD Slack](https://cloud-native.slack.com/archives/C01TSERG0KZ/p1638978466372600?thread_ts=1638965338.363300&cid=C01TSERG0KZ), an improvement on the message would help a lot I reckon.

**Comment by swarren83** [comment]
+1, spent 2 days trying to figure out why an Application was in Unknown state.

**Comment by mnacharov** [comment]
I did some investigation of this Issue

When we don't pass any auth to gitlab we have this output
```
$ git clone  git@gitlab.com:m.nacharov/argocd-issue-4441.git
Cloning into 'argocd-issue-4441'...
no such identity: /home/mikhail/.ssh/argocd-issue-4441: No such file or directory
git@gitlab.com: Permission denied (publickey).
fatal: Could not read from remote repository.
```

When we pass ssh-key which not added to deploy keys of cloned repository we got this
```
git clone  git@gitlab.com:m.nacharov/argocd-issue-4441.git
Cloning into 'argocd-issue-4441'...
remote: 
remote: ========================================================================
remote: 
remote: ERROR: The project you were looking for could not be found or you don't have permission to view it.

remote: 
remote: ========================================================================
remote: 
fatal: Could not read from remote repository.
```
As I understood only first line of git output passes to the Application Status

So, we have a choices here:

1. Just though away all `remote:` lines. 
   This way we will have the same message in app status as we have when forgot provide deploy key for repository in argocd-cm
2. Read all `remote:` lines and first line after them, remove all empty and `=====` lines
   This way we will get more specific, but a bit more verbose message at app status. But we need to have some gitlab-specific processing of remote messages in argo-cd source code

**Comment by mnacharov** [comment]
Looks like this error comes from github.com/go-git/go-git/v5 client used by [getRefs](https://github.com/argoproj/argo-cd/blob/9fac0f6ae6e52d6f4978a1eaaf51fbffb9c0958a/util/git/client.go#L431) method.
When I [provide commit sha](https://github.com/argoproj/argo-cd/blob/9fac0f6ae6e52d6f4978a1eaaf51fbffb9c0958a/util/git/client.go#L524-L525) I see good and verbose error message
![image](https://user-images.githubusercontent.com/1246847/189925191-cfb1a224-9e3e-4be8-b9d2-c6d0a53f0443.png)
